### ISSUE TITLE
Dropdown: How does the select value get updated? (AngularJS value update issue)

### ISSUE THREAD

**Issue by cbier** [Formstone/Formstone#129]
I am trying to figure out where you're setting the `<select>` value after a user chooses a value. So far I've discovered that `handleChange()` and `updateOption()` but it doesn't look like either changes the value of the `<select>` element.

How is the value being changed after the user selects a value from the formstone dropdown?

P.S.
I'm trying to figure this out to help solve this issue: When the user chooses an item in the dropdown by typing on the keyboard and then clicks somewhere else on the page, AngularJS does not see the changed value.

Thanks in advance, and thanks for this great library :)

**Comment by cbier** [comment]
Ok I think that I see it now. In `updateOptions()`:

`data.$el[0].selectedIndex = index;`

So the value gets updated the same way whether you select it by pressing a key or clicking, right?

**Comment by cbier** [comment]
I see that `handleChange()` triggers the `change` event. I also noticed that it is not called in `onKeyPress()` like it is in other places (e.g. `onSelect()`). 

When I add `handleChange()` to `onKeyPress()` my issue with Angular not recognizing the new value is resolved. I think that Angular was not picking up the change because the `change` event was not being triggered.

My update in `onKeyPress()`:

```
            // Update
            if (index >= 0) {
                updateOption(index, data);
                handleChange(data);
                scrollOptions(data);
            }
```

I am curious what your thoughts/insight is on this @benplum. Should `handleChange()` be added to the end of `onKeyPress()`?

**Comment by benplum** [comment]
This follows the browser behavior of only triggering the change event when focus is removed from the select element. You'll notice the same behavior if you test the select without the plugin applied.

**Comment by cbier** [comment]
Thanks for the response! That makes sense. I would rather stick to emulating the browser behavior.

For some reason it seems that when I filter using the keyboard and then click somewhere else on the page (remove focus from the dropdown) the change event isn't being fired.

```
    function onFocusOut(e, internal) {
        Functions.killEvent(e);

        var $target = $(e.currentTarget),
            data    = e.data;

        if (data.focused && data.closed) {
            data.focused = false;

            data.$dropdown.removeClass(RawClasses.focus)
                          .off(Events.keyDown + data.dotGuid);

            if (!data.multiple) {
                // Clean up
                closeOptions(data);
            }
        }
    }
```

I don't see it in this function, but where in the code does it trigger the change value when the focus is removed?

**Comment by benplum** [comment]
That looks like a bug, I'll push an update shortly. 

Digging deeper, it looks like the browser behavior varies between vendors. Formstone is following the Firefox model (https://bugzilla.mozilla.org/show_bug.cgi?id=126379) of firing the event on click or focus out (will fix), but would it make more sense to follow the WebKit/Blink model of firing the event on click or keyboard, as you initially expected? Looking at stats it would probably make more sense to follow the most popular browsers, but am open to input before making a change.

**Comment by cbier** [comment]
Ok thanks Ben! Firing the change event on focus out works for my implementation.

The only reason that I put `handleChange()` in `onKeyPress()` because I couldn't think of anything else to address my problem.

I think that focus out makes a bit more sense because sometimes people (like me) use JS to change the UI based on the current selection. So triggering the change event on every keypress might cause some trouble for developers who are looking to change the UI only once a final value is selected.

**Comment by benplum** [comment]
I completely agree - it would then require the developer to manually debounce the events or something. 

I'll should be able to get focusout fix pushed today.

**Comment by cbier** [comment]
Ben, thanks so much!

**Comment by benplum** [comment]
This is resolved in 0.8.34, let me know if you're still having issues with the change events when navigating with the keyboard.

**Comment by cbier** [comment]
Thanks, Ben! I'll let you know when I get a chance to incorporate the change. I have a modified version of dropdown.js locally so I have to merge it in manually
